src/email/outbound.js

```diff
@@ -0,0 +1,326 @@
+            data = parts[parts.length - 1];
+        });
+
+        sock.on('error', function (err) {
+            throw Error(`Failed to connect to ${domain}: ${err}`);
+        });
+
+        let data = '';
+        let step = 0;
+        let loginStep = 0;
+        const queue = [];
+        // const login = [];
+        let parts;
+        let cmd;
+        let upgraded = false;
+        let isUpgradeInProgress = false;
+
+        queue.push('MAIL FROM:<' + from + '>');
+        const recipientsLength = recipients.length;
+        for (let i = 0; i < recipientsLength; i++) {
+            queue.push('RCPT TO:<' + recipients[i] + '>');
+        }
+        queue.push('DATA');
+        queue.push('QUIT');
+        queue.push('');
+
+        function response(code, msg) {
+            switch (code) {
+                case 220:
+                    //220   on server ready
+                    if (isUpgradeInProgress === true) {
+                        sock.removeAllListeners('data');
+                        let original = sock;
+                        original.pause();
+                        let opts = {
+                            socket: sock,
+                            host: sock._host,
+                            rejectUnauthorized,
+                        };
+                        if (startTLS) {
+                            opts.secureContext = createSecureContext({ cert: tls.cert, key: tls.key });
+                        }
+                        sock = connect(
+                            opts,
+                            () => {
+                                sock.on('data', function (chunk) {
+                                    data += chunk;
+                                    parts = data.split(CRLF);
+                                    const partsLength = parts.length - 1;
+                                    for (let i = 0, len = partsLength; i < len; i++) {
+                                        onLine(parts[i]);
+                                    }
+                                    data = parts[parts.length - 1];
+                                });
+                                sock.removeAllListeners('close');
+                                sock.removeAllListeners('end');
+                            }
+                        );
+                        sock.on('error', function (err) {
+                            log.warn("Could not upgrade to TLS:", err, "Falling back to plaintext");
+                        });
+                        // Resume plaintext connection
+                        original.resume();
+                        upgraded = true;
+                        w("EHLO " + srcHost);
+                        break;
+                    }
+                    else {
+                        // check for ESMTP/ignore-case
+                        if (/\besmtp\b/i.test(msg)) {
+                            // TODO: determine AUTH type; auth login, auth crm-md5, auth plain
```

Comment:
Suggestion: For each of the TODOs you want to do later, make an issue in this repo and link back to it as a comment. It means any community member can:

1) See what doesn't work yet
2) See what they can work on to make it work better
3) Find the code line the issue relates to